src/main/java/com/hyperlocal/server/ShopController.java

```diff
@@ -1,64 +1,233 @@
   private static final String DELETE_CATALOG_STATEMENT = "DELETE FROM `Catalog` WHERE `ServiceID` = ?;";
   private static final String PUBSUB_URL = "projects/speedy-anthem-217710/topics/testTopic";
   private static final Logger logger = LogManager.getLogger(ShopController.class);
 
+  @Inject
   public ShopController(PubSubTemplate pubSubTemplate) {
     this.publisher = pubSubTemplate;
     connection = MySQLConnectionBuilder.createConnectionPool(DATABASE_URL);
   }
 
-  /*
-   * To-do server-side checks: - Access control - Data validation
-   */
+  // API for performing text search 
+  @GetMapping("/api/query/elastic")
+  public CompletableFuture<List<ShopDetails>> getDataFromElasticSearch(@RequestParam String query,
+      @RequestParam String queryRadius, @RequestParam String latitude, @RequestParam String longitude) {
+
+    List<Long> shopIDList = new ArrayList<Long>();
+
+    // Create a match query for text Match
+    MultiMatchQueryBuilder matchQuery = QueryBuilders
+        .multiMatchQuery(query, "shopname", "typeofservice", "merchantname", "catalogitems").fuzziness("AUTO");
+
+    // GeoDistance query for filtering everything in a radius
+    GeoDistanceQueryBuilder filterOnDistance = QueryBuilders.geoDistanceQuery("pin.location")
+        .point(Double.parseDouble(latitude), Double.parseDouble(longitude)).distance(queryRadius);
+
+    // Boolean query to ensure condition of both Matchquery and Geodistance query holds
+    BoolQueryBuilder boolMatchQueryWithDistanceFilter = QueryBuilders.boolQuery().must(matchQuery)
+        .filter(filterOnDistance);
+
+    // Create a search request with the Boolean query
+    SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder();
+    searchSourceBuilder.query(boolMatchQueryWithDistanceFilter);
+
+    // Create the HTTP Request to send
+    HttpClient client = HttpClient.newHttpClient();
+    HttpRequest request = HttpRequest.newBuilder().uri(URI.create(SEARCH_INDEX_URL))
+        .method("GET", HttpRequest.BodyPublishers.ofString(searchSourceBuilder.toString()))
+        .setHeader("Content-Type", "application/json").build();
+
+    // Send request to search Index asynchronously and parse response to get ShopIDs
+    return client.sendAsync(request, BodyHandlers.ofString()).thenApply(HttpResponse::body)
+        .thenApply((responseString) -> {
+          JsonObject obj = JsonParser.parseString(responseString).getAsJsonObject();
+          JsonArray idListJson = obj.get("hits").getAsJsonObject().get("hits").getAsJsonArray();
+          for (JsonElement id : idListJson) {
+            shopIDList.add(id.getAsJsonObject().get("_id").getAsLong());
+          }
+          return shopIDList;
+        
+        // Perform BatchQuery on shopIDs and get List of ShopDetails corresponding to the shopIDs
+        }).thenCompose((shopList) -> {
```

Comment:
In lieu of chaining a `thenCompose()` after the `thenApply()` method, you can get rid of `thenApply()` method altogether and call `getShopsByShopIDBatch()` in the same lambda function which prepares the `shopIDList`.

Reply:
Actually. Thanks for this!! Let me make the change at other places too